sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamRelNode.java

```diff
@@ -17,18 +17,35 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl.rel;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.Pipeline;
 import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.values.PCollection;
-import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PInput;
 import org.apache.beam.sdk.values.Row;
 import org.apache.calcite.rel.RelNode;
 
 /** A {@link RelNode} that can also give a {@link PTransform} that implements the expression. */
 public interface BeamRelNode extends RelNode {
 
-  /**
-   * A {@link BeamRelNode} is a recursive structure, the {@code BeamQueryPlanner} visits it with a
-   * DFS(Depth-First-Search) algorithm.
-   */
-  PTransform<PCollectionTuple, PCollection<Row>> toPTransform();
+  /** Transforms the inputs into a PInput. */
+  default PInput buildPInput(Pipeline pipeline, Map<Integer, PCollection<Row>> cache) {
+    List<RelNode> inputs = getInputs();
+    if (inputs.size() == 0) {
```

Comment:
I want to highlight that at the core model level, everything is a `PCollectionTuple`, more or less. Every `PInput.expand()` method converts to such a structure. So this is just an empty tuple. I think it is safe to keep the input type `PCollectionTuple`. Seems like it ought to simplify things?